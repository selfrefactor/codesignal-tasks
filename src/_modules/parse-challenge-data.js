import {
  head,
  last,
  map,
  piped,
  sort,
  split,
  trim,
} from 'rambdax'

const EXPECTED_OUTPUT = 'Expected Output:'
const EXPECTED_OUTPUT_END = 'Click the '

function whenNext(testInput, next, rawText){
  const sk = piped(
    rawText,
    split(next),
    head,
    split(`${ testInput }:`),
    last,
    trim
  )

  return sk
}

function whenLast(testInput, rawText){
  const sk = piped(
    rawText,
    split(EXPECTED_OUTPUT),
    head,
    split(`${ testInput }:`),
    last,
    trim
  )

  return sk
}

function parseTestCase(testCase, sortedTestInputs){
  const hash = {}

  sortedTestInputs.forEach((x, i) => {
    const next = sortedTestInputs[ i + 1 ]
    if (!next){
      return hash[ x ] = whenLast(x, testCase.rawText)
    }

    hash[ x ] = whenNext(x, next, testCase.rawText)
  })
  const expected = piped(
    testCase.rawText,
    split(EXPECTED_OUTPUT_END),
    head,
    split(EXPECTED_OUTPUT),
    last,
    trim
  )

  return {expected, testInputs: hash}
}

export function parseChallengeData(challengeData){
  const testInputs = piped(
    challengeData.replContent,
    split('('),
    last,
    split(')'),
    head,
    split(','),
    map(trim)
  )
  if (testInputs.length === 0) throw new Error('testInputs.length')

  const sortedTestInputs = sort((x, y) => {
    const xPosition = challengeData.testCases[0].rawText.indexOf(x)
    const yPosition = challengeData.testCases[0].rawText.indexOf(y)

    return xPosition < yPosition ? -1 : 1
  }, testInputs)

  const testCases = piped(
    challengeData.testCases,
    map(x => parseTestCase(x, sortedTestInputs))
  )

  return {testCases, functionName: challengeData.kataTitle, sortedTestInputs}
}
