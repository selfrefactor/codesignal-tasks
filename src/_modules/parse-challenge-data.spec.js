import { parseChallengeData } from './parse-challenge-data'
import {data} from '../challange-test-data.json'
import { writeJson } from 'fs-extra'

test('happy', async () => {
  const result = parseChallengeData(data)
  await writeJson(`${__dirname}/parsed-test-data.json`, {data:result})

  expect(
    result
  ).toMatchSnapshot()
})