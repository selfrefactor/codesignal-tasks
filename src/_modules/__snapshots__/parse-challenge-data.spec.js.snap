// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy 1`] = `
Object {
  "functionName": "minimalBasketPrice",
  "sortedTestInputs": Array [
    "maxPrice",
    "vendorsDelivery",
    "vendorsProducts",
  ],
  "testCases": Array [
    Object {
      "expected": "[1, 2]",
      "testInputs": Object {
        "maxPrice": "7",
        "vendorsDelivery": "[5, 4, 2, 3]",
        "vendorsProducts": "[[1,1,1], 
 [3,-1,3], 
 [-1,2,2], 
 [5,-1,-1]]",
      },
    },
    Object {
      "expected": "[1, 2, 3]",
      "testInputs": Object {
        "maxPrice": "6",
        "vendorsDelivery": "[1, 5, 10, 12]",
        "vendorsProducts": "[[-1,-1,-1], 
 [3,-1,-1], 
 [-1,2,-1], 
 [-1,-1,1]]",
      },
    },
    Object {
      "expected": "[0]",
      "testInputs": Object {
        "maxPrice": "5",
        "vendorsDelivery": "[5, 6]",
        "vendorsProducts": "[[5], 
 [6]]",
      },
    },
    Object {
      "expected": "[0]",
      "testInputs": Object {
        "maxPrice": "1000000",
        "vendorsDelivery": "[1000000]",
        "vendorsProducts": "[[1000000]]",
      },
    },
    Object {
      "expected": "[0, 2, 3]",
      "testInputs": Object {
        "maxPrice": "6",
        "vendorsDelivery": "[100, 5, 10, 12]",
        "vendorsProducts": "[[3,4,4], 
 [4,4,4], 
 [4,2,4], 
 [4,4,1]]",
      },
    },
    Object {
      "expected": "[0]",
      "testInputs": Object {
        "maxPrice": "10",
        "vendorsDelivery": "[1]",
        "vendorsProducts": "[[10]]",
      },
    },
  ],
}
`;
